app/validators/questionnaire_validator.py

```diff
@@ -134,3 +139,90 @@ def validate_introduction_block(self):
         )
         if not has_introduction_blocks:
             self.add_error(error_messages.PREVIEW_WITHOUT_INTRODUCTION_BLOCK)
+
+    def validate_list_references(self):
+        referenced_lists = self.populate_referenced_lists()
+
+        # We need to keep track of section index for: common_definitions.json#/section_enabled
+        for section_index, section in enumerate(self.questionnaire_schema.sections):
+            identifier_references = get_object_containing_key(section, "source")
+            for _, identifier_reference, parent_block in identifier_references:
+                if identifier_reference["source"] == "list":
+                    list_identifier = identifier_reference["identifier"]
+                    if parent_block:
+                        # Parent block might not always be top level "Block" so we need to resolve it below
+                        parent_block_index = self.questionnaire_schema.resolve_parent_block_index_for_source(
+                            parent_block["id"]
+                        )
+                        if (
+                            parent_block_index
+                            < referenced_lists[list_identifier]["block_index"]
+                        ):
+                            self.add_error(
+                                error_messages.LIST_REFERENCED_BEFORE_ADDED.format(
+                                    list_name=list_identifier
+                                ),
+                                section_name=section["id"],
+                            )
+                    elif (
+                        section_index
+                        < referenced_lists[list_identifier]["section_index"]
+                    ):
+                        # Section level "enabled" rule that can use list source,
+                        # check: common_definitions.json#/section_enabled
+                        self.add_error(
+                            error_messages.LIST_REFERENCED_BEFORE_ADDED.format(
+                                list_name=list_identifier
+                            ),
+                            section_name=section["id"],
+                        )
+
+    def populate_referenced_lists(self):
```

Comment:
Since this is just populating, I would move this to questionnaire_schema and populate the map up front similar to what we the way we do in runner

Reply:
I made it questionnaire schema `@property` similar to `answers_with_context`. Thanks.